ibis/backends/tests/test_impure.py

```diff
@@ -0,0 +1,207 @@
+            marks=[
+                *no_udfs,
+                pytest.mark.broken(
+                    ["flink", "postgres"], reason="instances are uncorrelated"
+                ),
+            ],
+            id="udf",
+        ),
+    ],
+)
+
+
+@pytest.mark.broken("sqlite", reason="instances are uncorrelated")
+@mark_impures
+def test_impure_correlated(alltypes, impure):
+    # An "impure" expression is random(), uuid(), or some other non-deterministic UDF.
+    # If we evaluate it for two different rows in the same relation,
+    # we might get different results. This is expected.
+    # But, as soon as we .select() it into a new relation, then that "locks in" the
+    # value, and any further references to it will be the same.
+    # eg if you look at the following SQL:
+    # WITH
+    #   t AS (SELECT random() AS common)
+    # SELECT common as x, common as y FROM t
+    # Then both x and y should have the same value.
+    df = (
+        alltypes.select(common=impure(alltypes))
+        .select(x=_.common, y=_.common)
+        .execute()
+    )
+    tm.assert_series_equal(df.x, df.y, check_names=False)
+
+
+@pytest.mark.broken("sqlite", reason="instances are uncorrelated")
+@mark_impures
+def test_chained_selections(alltypes, impure):
+    # https://github.com/ibis-project/ibis/issues/8921#issue-2234327722
+    # This is a slightly more complex version of test_impure_correlated.
+    # consider this SQL:
+    # WITH
+    #   t AS (SELECT random() AS num)
+    # SELECT num, num > 0.5 AS isbig FROM t
+    # We would expect that the value of num and isbig are consistent,
+    # since we "lock in" the value of num by selecting it into t.
+    t = alltypes.select(num=impure(alltypes))
+    t = t.mutate(isbig=(t.num > 0.5))
+    df = t.execute()
+    df["expected"] = df.num > 0.5
+    tm.assert_series_equal(df.isbig, df.expected, check_names=False)
+
+
+impure_params_uncorrelated = pytest.mark.parametrize(
+    "impure",
+    [
+        pytest.param(
+            lambda _: ibis.random(),
+            marks=[
+                *no_randoms,
+                pytest.mark.broken(
+                    ["impala", "trino"], reason="instances are correlated"
+                ),
+            ],
+            id="random",
+        ),
+        pytest.param(
+            # make this a float so we can compare to .5
+            lambda _: ibis.uuid().cast(str).contains("a").ifelse(1, 0),
+            marks=[
+                *no_uuids,
+                pytest.mark.broken(
+                    ["mysql", "trino"], reason="instances are correlated"
+                ),
+            ],
+            id="uuid",
+        ),
+        pytest.param(
+            lambda table: my_random(table.float_col),
+            marks=[
+                *no_udfs,
+                pytest.mark.broken("duckdb", reason="instances are correlated"),
+            ],
+            id="udf",
+        ),
+    ],
+)
+
+
+@pytest.mark.broken(["clickhouse"], reason="instances are correlated")
```

Comment:
Did you look into why this is?

Two calls to `rand` generate the same number:

```
falcon :) select rand32(), rand32();

SELECT
    rand32(),
    rand32()

Query id: edfb1ecd-8070-48d9-8fc7-1b49b6f46fd9

   ┌───rand32()─┬───rand32()─┐
1. │ 3907068806 │ 3907068806 │
   └────────────┴────────────┘

1 row in set. Elapsed: 0.001 sec.
```

Do we now need to start seeding these functions with a random number from Python to pass to clickhouse?

What if someone is share a clickhouse connection across threads, does Ibis now need to manage random number generate thread local state?

These are the kinds of problems that we have to consider before writing ourselves into a corner with test cases.

Reply:
Hmm, that is interesting. I don't think the SQL spec is clear here on what is correct behavior, and I suppose even if I didn't know SQL and I looked at `select rand(), rand();` I wouldn't be surprised if you told me they could eval to the same value.

What about relaxing this test to checking that `t.select(x=ibis.random()).select(y=ibis.random())` yields `x != y`?

I think this would leave us with ibis semantics of
- if in the same select statement, two pure functions WILL give the same results (duh)
- if in different select statements, two pure functions WILL give the same results (duh)
- if in the same select statement, two impure functions may or may not give the same results. Ibis makes no guarantees. IDK, someone might still be able to footgun themselves, but at least right now I can't think of a way of protecting them...
- if in different select statements, two impure functions WILL give different results.
- if any value (pure or impure) is defined in one select statement, then any references to it in subsequent .select()s WILL all give the same value